### ISSUE TITLE
general ranges for dashboards like mixpanel

### ISSUE THREAD

**Issue by adrianthedev** [avo-hq/avo#2301]

**Comment by Bergrebell** [comment]
this looks fantastic!

i was looking for a solution to something similar in my previous avo-project, where we wanted to show some figures per user (we have multiple charts which show different activities per user. the user can be selected in each chart via a select. but when we wanted to see the performance of a user, we had to go through all the charts and select the user in every single one of them). am i correct in assuming that this will allow us to select one user on the dashboard and adapt all of the charts?

**Comment by adrianthedev** [comment]
Yes, that is correct. All cards will be influenced by this range/filter.
We figured out this use-case a while back, but after the initial release.

**Comment by alexcwatt** [comment]
@adrianthedev I'm building a project this week with Avo and it would be sweet to have this feature. I am curious to see the state of the referenced PR and what might need to be done to finish it.

**Comment by adrianthedev** [comment]
I need to refresh my memory, but I think the back-end is pretty much done. Or at least good enough for the initial iteration, and all it needs is the front-end and the API to manage the filters.
Would you like to give this a go?
I can invite you to the private repo and even have @Paul-Bob help with onboarding so you can get started.

**Comment by jpowell** [comment]
I bet AI could help knock this out pretty quick... especially if what's left has reference implementations that can be tweaked slightly

**Comment by alexcwatt** [comment]
@adrianthedev Sure, I can't guarantee; I can work around the gap by building a custom report page and I very well might need that anyway. But if you add me, I can work on this if it ends up making the most sense for my use case. Thanks!

**Comment by alexcwatt** [comment]
@adrianthedev I've made good progress on other aspects of the project and would prefer to build the reporting and visualizations using an Avo dashboard than rolling something completely different. Do you mind adding me to the repo? I will see what I can do over the next couple days.

**Comment by Paul-Bob** [comment]
Hi @alexcwatt, you should already have access to the repo.

Let me know if you need any assistance with onboarding. If so, we can hop on a call, you can use [this link](https://savvycal.com/avo-hq/paul-support) to schedule it.

**Comment by jannisborgers** [comment]
@alexcwatt Were you able to make progress on the front end? It would be nice to cut time on our custom solution and switch to the official one instead.

Happy to have found this issue. We need to implement the same: A previous/next date selection and a date-picker where I can select a single day to look into for the data. Currently, I use a partial card that has the necessary form/links:

```erb
<% current_date = (params[:report_date] || Date.current.to_s).to_date %>

<a href="?report_date=<%= current_date - 1 %>" data-turbo="false">← Previous day</a>
<form method="get" data-turbo="false">
  <input
    type="date"
    name="report_date"
    value="<%= current_date %>"
    class="form-control"
  />
</form>
<a href="?report_date=<%= current_date + 1 %>" data-turbo="false">Next day →</a>

<a href="?report_date=<%= Date.current %>" data-turbo="false">Today</a>
```

And render it with the other cards:

```erb
class Avo::Dashboards::Report < Avo::Dashboards::BaseDashboard
  self.id = "report"
  self.name = "Report"

  def cards
    selected_date = Date.parse(Avo::Current.params[:report_date]) rescue Date.current

    card Avo::Cards::DateSelectorCard

    card Avo::Cards::SomeMetric, label: "First card", arguments: {
      some_arg: foo,
      date: selected_date
    }

    card Avo::Cards::SomeMetric, label: "Second card", arguments: {
      some_arg: bar,
      date: selected_date
    }

    <%# etc. %>
  end
end
```

Because the partial card gets rendered in a turbo frame, I set `data-turbo="false"` on the form and all links in order for the page to reload and `current_date` to update. 

But if the data could update without reloading, that would be a lot smoother.

**Comment by alexcwatt** [comment]
@jannisborgers unfortunately no, I don't have anything to help here. I will keep following the issue and would be happy to leverage something you build!

**Comment by jannisborgers** [comment]
@alexcwatt Thanks for the update. @adrianthedev Is there any way that you can think of that allows us to:

- add a partial card with custom forms to select filters (date ranges, model properties, etc)
- allow the card to send the new data to the turbo frame of the dashboard, but not to the top-level page
- refresh all cards based on new the new data in the dashboard?

I'm all for having a dashboard-wise range selector, but we quickly needed more filter inputs not related to just ranges. This would allow us to use partial cards for a custom filter UI, which updates everything.

Or would that have to be a tool instead, which would mean losing the cards feature?
